Update HV scan, and badROC exclusion during HV scan

bartokm opened the pull request.

bartokm committed:
Update HV scan, and badROC exclusion during HV scan

ferencek commented:
Is this PR part of integrating HV bias scan code and settings, which up to now went into [SiPixelTools-HVScanAnalysis](https://github.com/CMSTrackerDPG/SiPixelTools-HVScanAnalysis), into [SiPixelTools-PixelHistoMaker](https://github.com/CMSTrackerDPG/SiPixelTools-PixelHistoMaker) with the final goal of retiring [SiPixelTools-HVScanAnalysis](https://github.com/CMSTrackerDPG/SiPixelTools-HVScanAnalysis)? I remember we discussed this option at some point so just wanted double-check. I would guess there should also be some documentation that needs to be transferred from [SiPixelTools-HVScanAnalysis](https://github.com/CMSTrackerDPG/SiPixelTools-HVScanAnalysis).

bartokm commented:
Yes you're right, these are from https://github.com/CMSTrackerDPG/SiPixelTools-HVScanAnalysis
And if everything goes well we can retire that. I've tested it in a 2022 miniscan and L1 full scan, the results are the same except for hit efficiency. Which is somewhat expected, since the definition there changed. We need to discuss the hit efficiency part in a meeting at some point, but I think this is a good first step towards merging the 2 codes.

Documentation: yeah I'll update it, I just like it better to edit the readme text directly through the browser.

bartokm commented on Phase1ScanHistoMaker.cc:
```diff
@@ -507,6 +517,7 @@ int main(int argc, char* argv[]) {
   sh.AddHistos("clust", { .fill="ROC_Ring2_BladePanel_vs_Disk",                        .pfs={"DelayScans"},          .cuts={"ZeroBias","Ring2"}, .draw="COLZ", .opt="Log", .ranges={0,0, 0,0, 0,0} });
 #endif
   
+  std::function<void()> add_special_histos = [&sh]{};
```
@TizianoBevilacqua I had to move this since in the original the code broke if I switched to HV scan mode (#define HV_Scan     1). Please look at this special histos part if it's okay this way.

TizianoBevilacqua replied:
By just looking at it it seems fine, have you tested it?

ferencek commented:
@bartokm, @TizianoBevilacqua, if this is not interfering with anything, I would propose to merge this. Otherwise, we risk getting merge conflicts if other developments happen in the meantime.

TizianoBevilacqua commented on interface/Variables.h:
```diff
@@ -1926,7 +1972,7 @@ class Variables {
     effcut_dz          =       cut_nvtx && cut_federr && hp && pt && nstrip && d0 &&       pixhit && noscan && goodmod && goodroc && lx_fid && ly_fid && valmis && hitsep && trk_beta_cut;
     effcut_d0_dz       =       cut_nvtx && cut_federr && hp && pt && nstrip &&             pixhit && noscan && goodmod && goodroc && lx_fid && ly_fid && valmis && hitsep && trk_beta_cut;
-    effcut_scans       =       cut_nvtx && cut_federr && hp && pt && nstrip && d0 && dz && pixhit           && goodmod            && lx_fid && ly_fid && valmis && hitsep && trk_beta_cut;
+    effcut_scans       =       cut_nvtx && cut_federr && hp && pt && nstrip && d0 && dz && pixhit           && goodmod && goodroc_scan && lx_fid && ly_fid && valmis && hitsep && trk_beta_cut;
```
I'm not sure I understand this change correctly, is it needed just for the HV scan right? If the run is not a HV scan in line 1464 nothing should change right? because this cut affects also the Timing scan

bartokm replied:
This is part of a more complex modification. So badroc exclusion for scans (HV and Timing) are changed to be the following:
1. Use a specific run for the exclusion defined in scan_points.h
2. If no run is defined, use the same run, but excluding the scan points when HV is changing.
So basically since for timing scan this is not relevant, it should be transparent for that. But maybe it should be checked.

bartokm commented:
It's okay by me. I just had 1 question to Tiziano, because I've modified some of his stuff.
I'll write a readme for the HV scan later...

bartokm commented:
I don't use this special histos, I've only checked for HV scans, it works there.

ferencek commented:
Hi @TizianoBevilacqua, @bartokm, we somehow dropped the ball on this. I think we agreed that Tiziano would re-run this year's timing scan (the full timing scan was presented [here](https://indico.cern.ch/event/1271485/#17-results-from-apr-21-full-ti)) with this PR locally merged to see if the scan results come out unchanged. If that is confirmed, we can then merge this PR. Marton, you also mentioned updating the ReadMe file once the PR is merged. Things have been quiet on the pixel offline front since the premature end of pp data taking and I can imagine that you both have moved on to other stuff that is of higher priority at the moment but it would be really nice to have this merged before the end of the year. What do you think?

Also keeping @mroguljic in the loop.

bartokm committed:
update readme with HVBiasScan info

bartokm commented:
There's one conflict we need to resolve: the beginning this year I switched to using this version of the PHM. But the hit efficiency plots were different w.r.t. 2022. I've identified a bunch of cuts which were tightened for hit eff. in this version. So I've modified those cuts to get the same old results. This PR is without that modification. If I include it, it would change the hit eff plots for the timing scan as well... One solution is to make the changes in a way that it only affects the HV scan. It's a bit of work, but I think I can do it.

ferencek commented:
If you can introduce a switch that controls cuts for different types of scans, feel free to add this change to this PR.

bartokm committed:
add 2023 scans, and some updates

bartokm commented:
I've made the necessary modifications. Timing scans should be unaffected this way. I've tested locally for the HV scan, it gives the same result.

ferencek commented:
Hi @bartokm, I am looking at changes in `Phase1ScanHistoMaker.cc` and I wonder why there are so many changes of `"HVBiasScans"` to `"HVBiasScans_2023"` even in parts that seem to be Phase 0-specific. In general, is it not better to use a more general name than to have year in the names of various variables which presumably will also be used for years other than 2023.

bartokm commented:
I just used search & replace that's why it's changed everywhere. I can get rid of the Phase0 stuff. But I doubt that anyone will use that part of the code again.
It is extremely annoying and very error prone to use such a long postfix (HVBiasScans https://github.com/CMSTrackerDPG/SiPixelTools-PixelHistoMaker/pull/2/files#diff-97f1bace7102d0b376be14f8f09bf8645ea8c856ef8cf402e25e8ffacdead6aeR234) that includes all scans done so far (and will include all scans in the future, if still people will use this code). That's why I introduced a yearly postfix, which needs to be changed each year. Let me know if you've a better suggestion.

ferencek commented:
For Phase 0 stuff I leave it up to you to decide but you are probably right that this is safe to remove. I agree that these long strings are error prone. I see that you have `HVBiasScans_2015`, `HVBiasScans_2016`, etc. so I would have the most recent set of scans simply called `HVBiasScans` so that `HVBiasScans_2023` does not have to appear in so many places in the code. The one with the longest string that is called `HVBiasScans` could be renamed to, for example, `HVBiasScans_all` or something likes that. So the idea would be that the plot for 2023 is called `HVBiasScans` and at some later point `HVBiasScans_2023` is created with the string that corresponds to 2023 scans and `HVBiasScans`  becomes a 2024 scan. That way the PR would not change `HVBiasScans` to `HVBiasScans_2023` in so many places.

TizianoBevilacqua committed:
adding changes to fix problems with PR2 for the TimingScan

bartokm committed:
change HV_Scan and Timing_Scan option to be set at compile

bartokm committed:
Change postfix HVBiasScans to be standard

bartokm committed:
Update README.md
